vmware-base/src/com/cloud/hypervisor/vmware/util/VmwareContextPool.java

```diff
@@ -45,76 +43,74 @@ public VmwareContextPool() {
 
         _timer.scheduleAtFixedRate(getTimerTask(), _idleCheckIntervalMs, _idleCheckIntervalMs);
     }
 
-    public void registerOutstandingContext(VmwareContext context) {
-        assert (context != null);
-        synchronized (this) {
-            _outstandingRegistry.add(context);
-        }
-    }
-
-    public void unregisterOutstandingContext(VmwareContext context) {
-        assert (context != null);
-        synchronized (this) {
-            _outstandingRegistry.remove(context);
+    public VmwareContext getContext(final String vCenterAddress, final String vCenterUserName) {
+        final String poolKey = composePoolKey(vCenterAddress, vCenterUserName);
+        if (Strings.isNullOrEmpty(poolKey)) {
+            return null;
         }
-    }
-
-    public VmwareContext getContext(String vCenterAddress, String vCenterUserName) {
-        String poolKey = composePoolKey(vCenterAddress, vCenterUserName);
         synchronized (this) {
-            List<VmwareContext> l = _pool.get(poolKey);
-            if (l == null)
-                return null;
-
-            if (l.size() > 0) {
-                VmwareContext context = l.remove(0);
-                context.setPoolInfo(this, poolKey);
-
-                if (s_logger.isTraceEnabled())
-                    s_logger.trace("Return a VmwareContext from the idle pool: " + poolKey + ". current pool size: " + l.size() + ", outstanding count: " +
-                        VmwareContext.getOutstandingContextCount());
+            final Queue<VmwareContext> ctxList = _pool.get(poolKey);
+            if (ctxList != null && ctxList.size() > 0) {
+                final VmwareContext context = ctxList.remove();
+                if (context != null) {
+                    context.setPoolInfo(this, poolKey);
+                }
+                if (s_logger.isTraceEnabled()) {
+                    s_logger.trace("Return a VmwareContext from the idle pool: " + poolKey + ". current pool size: " + ctxList.size() + ", outstanding count: " +
+                            VmwareContext.getOutstandingContextCount());
+                }
                 return context;
             }
-
-            // TODO, we need to control the maximum number of outstanding VmwareContext object in the future
             return null;
         }
     }
 
-    public void returnContext(VmwareContext context) {
+    public void registerContext(final VmwareContext context) {
         assert (context.getPool() == this);
         assert (context.getPoolKey() != null);
         synchronized (this) {
-            List<VmwareContext> l = _pool.get(context.getPoolKey());
-            if (l == null) {
-                l = new ArrayList<VmwareContext>();
-                _pool.put(context.getPoolKey(), l);
+            Queue<VmwareContext> ctxList = _pool.get(context.getPoolKey());
+            if (ctxList == null) {
+                ctxList = EvictingQueue.create(_maxIdleQueueLength);
```

Comment:
Since all access is in synchronized blocks this is not required. But a synchronised evictingqueue could have let you avoid some of synchronised blocks.

Again, this is not required as it is.